jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c

```diff
@@ -2661,9 +2663,217 @@ ecma_builtin_array_prototype_includes (const ecma_value_t args[], /**< arguments
+    }
+    bool shouldFlatten = false;
+    if (depth > 0)
+    {
+      ecma_value_t is_array = ecma_is_value_array (element);
+
+      if (ECMA_IS_VALUE_ERROR (is_array))
+      {
+        ecma_free_value (element);
+        return is_array;
+      }
+      shouldFlatten = ecma_is_value_true (is_array);
+    }
+
+    if (shouldFlatten)
+    {
+      ecma_object_t *element_obj = ecma_get_object_from_value (element);
+      uint32_t element_len;
+      ecma_value_t len_value = ecma_op_object_get_length (element_obj, &element_len);
+
+      if (ECMA_IS_VALUE_ERROR (len_value))
+      {
+        ecma_deref_object (element_obj);
+        return len_value;
+      }
+
+      ecma_value_t target_index_val = ecma_builtin_array_flatten_into_array (target,
+                                                                             element_obj,
+                                                                             element_len,
+                                                                             target_index,
+                                                                             depth-1,
+                                                                             ECMA_VALUE_UNDEFINED,
+                                                                             ECMA_VALUE_UNDEFINED);
+
+      ecma_deref_object (element_obj);
+
+      if (ECMA_IS_VALUE_ERROR (target_index_val))
+      {
+        return target_index_val;
+      }
+      target_index = (uint32_t) ecma_get_number_from_value (target_index_val);
+    }
+    else
+    {
+      uint8_t propertys = ECMA_PROPERTY_CONFIGURABLE_ENUMERABLE_WRITABLE;
+      ecma_value_t element_temp = ecma_builtin_helper_def_prop_by_index (ecma_get_object_from_value (target),
+                                                                         target_index,
+                                                                         element,
+                                                                         propertys);
+
+      ecma_free_value (element);
+      if (ECMA_IS_VALUE_ERROR (element_temp))
+      {
+        return element_temp;
+      }
+
+      target_index++;
+    }
+  }
+  return ecma_make_uint32_value (target_index);
+} /* ecma_builtin_array_flatten_into_array */
+
+/**
+ * The Array.prototype object's 'flat' routine
+ *
+ * See also:
+ *          ECMA-262 v10, 22.1.3.10
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_array_prototype_object_flat (const ecma_value_t args[], /**< arguments list */
+                                          ecma_length_t args_number, /**< number of arguments */
+                                          ecma_object_t *obj_p, /**< array object */
+                                          uint32_t len) /**< array object's length */
+{
+  ecma_number_t depthNum = 1;
+  if (args_number > 0 || ECMA_IS_VALUE_ERROR (ecma_op_to_integer (args[0], &depthNum)))
```

Comment:
That's  definitely incorrect. `[1,2].flat(5)` would raises an error here, and `[1,2].flat()` leads to invalid array element access.